### PR TITLE
Various improvements to the heartbeat

### PR DESCRIPTION
**Pull request by natefoo**
- Dump on `SIGUSR1`
- Configurable dump period (including "no periodic dump, signal only")
- Default log filename now includes `server_name` so logs do not clobber eachother
- Configurable/templateable log filename
- Catch more sleeping threads
- Use `write()` instead of `print`

Regarding `SIGUSR1`, this is only one action that Galaxy might take on `SIGUSR1`. Others could be added later. That said, the utility of Galaxy signal handling is somewhat limited by uWSGI claiming every conceivable signal.

### CODE REVIEW

File: lib/galaxy/app.py [hunk 1]
```diff
@@ -127,15 +136,25 @@ def __init__( self, **kwargs ):
             self.openid_providers = OpenIDProviders.from_file( self.config.openid_config_file )
         else:
             self.openid_providers = OpenIDProviders()
-        # Start the heartbeat process if configured and available
         from galaxy import auth
         self.auth_manager = auth.AuthManager( self )
+        # Start the heartbeat process if configured and available (wait until
+        # postfork if using uWSGI)
         if self.config.use_heartbeat:
             from galaxy.util import heartbeat
```
Comment (nsoranzo) [inline lib/galaxy/app.py]: Can the 2 `from ... import`s above be moved to the top of the file?

**Comment by jmchilton** [comment]
@natefoo Is there a good reason to have the default "off with a period of 20" instead of "on with a period of -1"?

**Comment by natefoo** [comment]
"On" starts an extra thread, so I'd prefer to leave it disabled.

**Comment by jmchilton** [comment]
Meh - threads are so lightweight I would not prevent the extra convince just to save one if it were me - but they are your changes so you can have my :+1:.

**Comment by natefoo** [comment]
"Less is better" ;)

File: lib/galaxy/app.py (hunk 1 above)
Comment (natefoo) [inline lib/galaxy/app.py]: There are others that are imported this way, seems to be mostly historic? It might've had to do with dependencies.

Comment (natefoo) [inline lib/galaxy/app.py]: Could also have been a timing issue with cyclical imports

File: lib/galaxy/app.py [hunk 2]
```diff
@@ -127,15 +137,24 @@ def __init__( self, **kwargs ):
             self.openid_providers = OpenIDProviders.from_file( self.config.openid_config_file )
         else:
             self.openid_providers = OpenIDProviders()
-        # Start the heartbeat process if configured and available
         from galaxy import auth
```
Comment (nsoranzo) [inline lib/galaxy/app.py:140]: Here's another import that can be moved if you want to clean it up.

Comment (natefoo) [inline lib/galaxy/app.py:140]: I'd prefer to do that in a separate PR
